pyo3-ffi/src/datetime.rs

```diff
@@ -0,0 +1,561 @@
+
+    #[link_name = "PyPyDateTime_DELTA_GET_DAYS"]
+    pub fn PyDateTime_DELTA_GET_DAYS(o: *mut PyObject) -> c_int;
+    #[link_name = "PyPyDateTime_DELTA_GET_SECONDS"]
+    pub fn PyDateTime_DELTA_GET_SECONDS(o: *mut PyObject) -> c_int;
+    #[link_name = "PyPyDateTime_DELTA_GET_MICROSECONDS"]
+    pub fn PyDateTime_DELTA_GET_MICROSECONDS(o: *mut PyObject) -> c_int;
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct PyDateTime_CAPI {
+    pub DateType: *mut PyTypeObject,
+    pub DateTimeType: *mut PyTypeObject,
+    pub TimeType: *mut PyTypeObject,
+    pub DeltaType: *mut PyTypeObject,
+    pub TZInfoType: *mut PyTypeObject,
+    #[cfg(not(all(PyPy, not(Py_3_8))))]
+    pub TimeZone_UTC: *mut PyObject,
+    pub Date_FromDate: unsafe extern "C" fn(
+        year: c_int,
+        month: c_int,
+        day: c_int,
+        cls: *mut PyTypeObject,
+    ) -> *mut PyObject,
+    pub DateTime_FromDateAndTime: unsafe extern "C" fn(
+        year: c_int,
+        month: c_int,
+        day: c_int,
+        hour: c_int,
+        minute: c_int,
+        second: c_int,
+        microsecond: c_int,
+        tzinfo: *mut PyObject,
+        cls: *mut PyTypeObject,
+    ) -> *mut PyObject,
+    pub Time_FromTime: unsafe extern "C" fn(
+        hour: c_int,
+        minute: c_int,
+        second: c_int,
+        microsecond: c_int,
+        tzinfo: *mut PyObject,
+        cls: *mut PyTypeObject,
+    ) -> *mut PyObject,
+    pub Delta_FromDelta: unsafe extern "C" fn(
+        days: c_int,
+        seconds: c_int,
+        microseconds: c_int,
+        normalize: c_int,
+        cls: *mut PyTypeObject,
+    ) -> *mut PyObject,
+    #[cfg(not(all(PyPy, not(Py_3_8))))]
+    pub TimeZone_FromTimeZone:
+        unsafe extern "C" fn(offset: *mut PyObject, name: *mut PyObject) -> *mut PyObject,
+
+    pub DateTime_FromTimestamp: unsafe extern "C" fn(
+        cls: *mut PyTypeObject,
+        args: *mut PyObject,
+        kwargs: *mut PyObject,
+    ) -> *mut PyObject,
+    pub Date_FromTimestamp:
+        unsafe extern "C" fn(cls: *mut PyTypeObject, args: *mut PyObject) -> *mut PyObject,
+    #[cfg(not(PyPy))]
+    pub DateTime_FromDateAndTimeAndFold: unsafe extern "C" fn(
+        year: c_int,
+        month: c_int,
+        day: c_int,
+        hour: c_int,
+        minute: c_int,
+        second: c_int,
+        microsecond: c_int,
+        tzinfo: *mut PyObject,
+        fold: c_int,
+        cls: *mut PyTypeObject,
+    ) -> *mut PyObject,
+    #[cfg(not(PyPy))]
+    pub Time_FromTimeAndFold: unsafe extern "C" fn(
+        hour: c_int,
+        minute: c_int,
+        second: c_int,
+        microsecond: c_int,
+        tzinfo: *mut PyObject,
+        fold: c_int,
+        cls: *mut PyTypeObject,
+    ) -> *mut PyObject,
+}
+
+// Python already shares this object between threads, so it's no more evil for us to do it too!
+unsafe impl Sync for PyDateTime_CAPI {}
+
+pub static mut PyDateTimeAPI: *mut PyDateTime_CAPI = ptr::null_mut();
```

Comment:
I ended up using a wrapped around Cell (since pointers are copy). When using pyo3_ffi directly I was exactly missing this kind of type for a similar purpose (for example initializing numpy api). Therefore I exposed this cell type as `UnsafeGILCell`. While this is conceptually a very simple type I think it makes writing unsafe ffi code in rust must more ergonomic (this is much better than mutable statics)